### ISSUE TITLE
Promise.finally should accept a function that returns Promise<void>

### ISSUE THREAD

**Issue by bogdanb** [microsoft/TypeScript#51029]
# lib Update Request

## Configuration Check

My compilation *target* is `ES2020` and my *lib* is `the default`.

## Missing / Incorrect Definition

`Promise.finally` is declared as `finally(onfinally?: (() => void) | undefined | null): Promise<T>`

It should instead be something like:

```
finally(onfinally?: (() => void | Promise<void>) | undefined | null): Promise<T>
```

It is perfectly valid to pass an `async` function to it, and the resulting promise will wait for `onfinally` to settle, before resolving with the original value. (Also, if `onfinally` throws or rejects, it overrides everything, but that’s irrelevant to the typing.)

## Sample Code

```
async function firstAfterBoth<A, B>(first: Promise<A>, second: Promise<B>): Promise<A> {
  return first.finally(async () => await second);
}
```

Typescript itself does not complain about it, but due to how the function is declared it triggers ESLint’s  `@typescript-eslint/no-misused-promises` rule (_Promise returned in function argument where a void return was expected_).

## Documentation Link

[MDN|https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally]
[ECMAScript Spec|https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-promise.prototype.finally]

**Comment by fatcerberus** [comment]
The `void` return itself feels problematic since a `void` return means “this function’s return value isn’t observed”, which is clearly not the case here—the return value *is* observed and if it’s a promise, the promise is awaited.

**Comment by MartinJohns** [comment]
Duplicate of #44980.
